scene/main/scene_tree.cpp

```diff
@@ -923,6 +923,59 @@ bool SceneTree::is_paused() const {
 	return paused;
 }
 
+int SceneTree::break_reference_cycles(bool p_scripts_only, bool p_collect_containers, bool p_verbose) {
+	static uint32_t pass = 1;
+	pass++;
+	tag_collect_pass(pass, p_collect_containers);
+
+	List<Engine::Singleton> singletons;
+	Engine::get_singleton()->get_singletons(&singletons);
+	for (const Engine::Singleton &S : singletons) {
+		S.ptr->tag_collect_pass(pass, p_collect_containers);
+	}
+	if (root) {
+		root->tag_collect_pass(pass, p_collect_containers);
+	}
+
+	List<ObjectID> collect_unlinked;
+	ObjectDB::fetch_unliked_objects(pass, collect_unlinked);
+
+	int unlinked_total = 0;
+
+	while (collect_unlinked.size()) {
+		Object *obj = ObjectDB::get_instance(collect_unlinked.front()->get());
+		if (obj) {
+			bool unlinked = false;
+			if (obj->get_script_instance()) {
+				if (p_verbose) {
+					print_line("Breaking cycle on: ", obj->to_string());
+				}
+
+				Ref<RefCounted> ref;
+				if (obj->is_ref_counted()) {
+					ref = Ref<RefCounted>(Object::cast_to<RefCounted>(obj)); // Keep a reference while breaking the link as this can cause a crash otherwise.
+				}
+				obj->set_script(Ref<Script>());
+				ref.unref();
+				unlinked = true;
+			}
+
+			if (!p_scripts_only) {
+				obj->unlink_resources(p_collect_containers);
+				unlinked = true;
+			}
+
+			if (unlinked) {
+				unlinked_total++;
+			}
+		}
+
+		collect_unlinked.pop_front();
+	}
```

Comment:
This is not about the code itself but the fact that we could use an "early exit" instead. I think we should promote this more often, as it would heavily reduce pyramid code.

```suggestion
	while (collect_unlinked.size()) {
		Object *obj = ObjectDB::get_instance(collect_unlinked.front()->get());
		if (!obj) {
		        continue;
		}
			
		bool unlinked = false;
		if (obj->get_script_instance()) {
			if (p_verbose) {
				print_line("Breaking cycle on: ", obj->to_string());
			}

			Ref<RefCounted> ref;
			if (obj->is_ref_counted()) {
				ref = Ref<RefCounted>(Object::cast_to<RefCounted>(obj)); // Keep a reference while breaking the link as this can cause a crash otherwise.
			}
			obj->set_script(Ref<Script>());
			ref.unref();
			unlinked = true;
		}

		if (!p_scripts_only) {
			obj->unlink_resources(p_collect_containers);
			unlinked = true;
		}

		if (unlinked) {
			unlinked_total++;
		}

		collect_unlinked.pop_front();
	}
```